apps/rich-list-api/src/modules/QueueModule.ts

```diff
@@ -0,0 +1,82 @@
+import { Module, Injectable } from '@nestjs/common'
+import { TypeOrmModule, InjectRepository } from '@nestjs/typeorm'
+import { Repository, PrimaryGeneratedColumn, Column, Entity, Index, Unique, In } from 'typeorm'
+
+export type Mode = 'FIFO' | 'LIFO'
+
+@Entity()
+@Unique(['queueName', 'data']) // ensure DE-DUPE built in, speed up find and replace
+@Index(['queueName', 'timestamp']) // for FIFO or LIFO purpose
+export class QueueItem {
+  @PrimaryGeneratedColumn()
+  id!: string
+
+  @Column()
+  data!: string
+
+  @Column()
+  queueName!: string
+
+  @Column({ type: 'bigint' })
+  timestamp!: number
+}
+
+@Injectable()
+export class QueueService {
+  constructor (
+    @InjectRepository(QueueItem)
+    private readonly repo: Repository<QueueItem>
+  ) {}
+
+  async createQueueIfNotExist (name: string, mode: Mode): Promise<Queue> {
+    return new Queue(name, mode, this.repo)
+  }
+}
+
+export class Queue {
+  constructor (
+    private readonly queueName: string,
+    private readonly mode: Mode,
+    private readonly repo: Repository<QueueItem>
+  ) {}
+
+  async receive (maxMessage: number): Promise<string[]> {
+    const consumed = await this.repo.find({
+      where: {
+        queueName: this.queueName
+      },
+      take: maxMessage,
+      order: {
+        timestamp: this.mode === 'FIFO' ? 'ASC' : 'DESC'
+      }
+    })
+    await this.repo.delete({ id: In(consumed.map(i => i.id)) })
```

Comment:
Wanted to ask a quick question considering the case of horizontal scalability. Are these two operations part of a single `transaction`? If two or more nodes are pulling from a queue before the delete operation we will have multiple nodes processing the same items of the queue.

These considerations may be moot, and I don't have enough background knowledge to determine the strictness of the queue. But just wanted to drop a quick question on this in case it could be an actual issue.